File: platform/test/connection/jira/fetch-httpx/test_case.py
#!/usr/bin/env python3
"""
Connection test for Jira API via the internal fetch_httpx package.

Tests proxy, verify_ssl, ca_bundle, and client cert configuration
by resolving config from AppYamlConfig and making a health-check request
using fetch_httpx.AsyncClient directly.

Auth: Basic auth with base64(email:api_token) per Atlassian REST API v3.
Health-check endpoint: GET /rest/api/3/myself
"""

from __future__ import annotations

import argparse
import asyncio
import base64
import json
import os
import re
import sys
import time
from pathlib import Path

_HERE = Path(__file__).resolve().parent
_PLATFORM = _HERE
while _PLATFORM.name != "platform" and _PLATFORM != _PLATFORM.parent:
    _PLATFORM = _PLATFORM.parent


def _inject_paths():
    """Add polyglot/packages_py source dirs to sys.path from platform pyproject.toml."""
    toml = _PLATFORM / "pyproject.toml"
    if not toml.exists():
        return
    for line in toml.read_text().splitlines():
        m = re.search(r'path\s*=\s*"([^"]+)"', line)
        if not m:
            continue
        pkg_path = (_PLATFORM / m.group(1)).resolve()
        src = pkg_path / "src"
        sys.path.insert(0, str(src if src.is_dir() else pkg_path))


_inject_paths()

from app_yaml_load import load_app_yaml_config
from fetch_httpx import AsyncClient, Proxy


# ---------------------------------------------------------------------------
# Connection config resolution
# ---------------------------------------------------------------------------

def _resolve_connection_config(provider_cfg: dict) -> dict:
    """Resolve proxy, SSL, and cert settings from YAML + env."""
    return {
        "proxy_url": (
            provider_cfg.get("proxy_url")
            or os.environ.get("HTTPS_PROXY")
            or os.environ.get("HTTP_PROXY")
        ),
        "verify_ssl": provider_cfg.get("verify_ssl", True),
        "ca_bundle": (
            provider_cfg.get("ca_bundle")
            or os.environ.get("SSL_CERT_FILE")
            or os.environ.get("REQUESTS_CA_BUNDLE")
        ),
        "client_cert": provider_cfg.get("client_cert"),
        "client_key": provider_cfg.get("client_key"),
    }


async def _test_connection(provider_cfg: dict, conn_cfg: dict) -> dict:
    """Test Jira API connection using fetch_httpx.AsyncClient directly."""
    email = (
        provider_cfg.get("email")
        or os.environ.get("JIRA_EMAIL")
    )
    api_token = (
        provider_cfg.get("api_token")
        or os.environ.get("JIRA_API_TOKEN")
    )
    base_url = (
        provider_cfg.get("base_url")
        or os.environ.get("JIRA_BASE_URL")
    )

    if not email:
        return {"skipped": True, "reason": "no JIRA_EMAIL or email in config"}
    if not api_token:
        return {"skipped": True, "reason": "no JIRA_API_TOKEN or api_token in config"}
    if not base_url:
        return {"skipped": True, "reason": "no JIRA_BASE_URL or base_url in config"}

    credentials = base64.b64encode(f"{email}:{api_token}".encode()).decode()

    proxy_url = conn_cfg["proxy_url"]
    verify_ssl = conn_cfg["verify_ssl"]
    ca_bundle = conn_cfg["ca_bundle"]
    client_cert = conn_cfg["client_cert"]
    client_key = conn_cfg["client_key"]

    proxy = Proxy(url=proxy_url) if proxy_url else None
    verify = ca_bundle if ca_bundle else verify_ssl
    cert = (client_cert, client_key) if client_cert else None

    try:
        start = time.monotonic()
        async with AsyncClient(
            base_url=base_url.rstrip("/"),
            headers={
                "Authorization": f"Basic {credentials}",
                "Accept": "application/json",
                "Content-Type": "application/json",
            },
            verify=verify,
            cert=cert,
            proxies=proxy,
            trust_env=True,
            timeout=15,
        ) as client:
            resp = await client.get("/rest/api/3/myself")
            latency = round((time.monotonic() - start) * 1000)
            data = resp.json() if resp.content else {}
            account_id = data.get("accountId") if isinstance(data, dict) else None
            display_name = data.get("displayName") if isinstance(data, dict) else None
            return {
                "skipped": False,
                "connected": resp.status_code < 400,
                "status_code": resp.status_code,
                "endpoint": "/rest/api/3/myself",
                "has_account_id": bool(account_id),
                "display_name": display_name,
                "latency_ms": latency,
                "connection": {
                    "base_url": base_url,
                    "proxy": proxy_url or "(none)",
                    "verify_ssl": verify_ssl,
                    "ca_bundle": ca_bundle or "(default)",
                    "has_client_cert": bool(client_cert),
                },
            }
    except Exception as exc:
        return {
            "skipped": False,
            "connected": False,
            "error": {"name": type(exc).__name__, "message": str(exc)},
        }


async def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--app-env", default="dev")
    parser.add_argument("--config-dir", required=True)
    args = parser.parse_args()

    app_cfg = load_app_yaml_config(config_dir=args.config_dir, app_env=args.app_env).config
    provider_cfg = app_cfg.get_nested("providers", "jira", default={}) or {}
    conn_cfg = _resolve_connection_config(provider_cfg)

    connection_test = await _test_connection(provider_cfg, conn_cfg)

    summary = {
        "service": "jira-fetch-httpx",
        "mode": "connection-test",
        "python_package": "fetch_httpx",
        "app_yaml_initialized": True,
        "connection_config": conn_cfg,
        "connection_test": connection_test,
        "overall": "PASS" if connection_test.get("connected") else (
            "SKIP" if connection_test.get("skipped") else "FAIL"
        ),
    }
    print(json.dumps(summary, indent=2, default=str))
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: platform/test/connection/jira/fetch-undici/test_case.mjs
#!/usr/bin/env node

/**
 * Connection test for Jira API via the internal fetch_undici package.
 *
 * Tests proxy, verify_ssl, ca_bundle, and client cert configuration
 * by resolving config from AppYamlConfig and making a health-check request
 * using AsyncClient with TLSConfig and Proxy support.
 *
 * Auth: Basic auth (email + API token).
 * Health-check endpoint: {baseUrl}/rest/api/3/myself
 */

import fs from 'node:fs';
import path from 'node:path';
import { pathToFileURL } from 'node:url';

const args = process.argv.slice(2);
const readArg = (name, fallback) => {
  const idx = args.indexOf(name);
  return idx >= 0 ? args[idx + 1] : fallback;
};

const appEnv = readArg('--app-env', process.env.APP_ENV || 'dev');
const configDir = readArg('--config-dir', process.env.CONFIG_DIR);
if (!configDir) {
  console.error('Missing --config-dir');
  process.exit(2);
}

const platformRoot = path.resolve(configDir, '..', '..');
const importFile = async (absPath) => import(pathToFileURL(absPath).href);

const appYamlLoadMod = await importFile(path.join(platformRoot, 'polyglot/app_yaml_load/mjs/dist/index.js'));
const fetchUndiciMod = await importFile(path.join(platformRoot, 'packages_mjs/fetch_undici/dist/index.js'));

const { config: appConfig } = await appYamlLoadMod.loadAppYamlConfig({ configDir, appEnv });
const providerConfig = appConfig.getNested(['providers', 'jira'], {}) || {};

// ---------------------------------------------------------------------------
// Connection config resolution
// ---------------------------------------------------------------------------

const connCfg = {
  proxyUrl: providerConfig.proxy_url || process.env.HTTPS_PROXY || process.env.HTTP_PROXY || null,
  verifySsl: providerConfig.verify_ssl ?? true,
  caBundlePath: providerConfig.ca_bundle || process.env.SSL_CERT_FILE || process.env.NODE_EXTRA_CA_CERTS || null,
  clientCert: providerConfig.client_cert || null,
  clientKey: providerConfig.client_key || null,
};

// ---------------------------------------------------------------------------
// Connection test
// ---------------------------------------------------------------------------

const email =
  providerConfig.email ||
  process.env.JIRA_EMAIL;

const apiToken =
  providerConfig.api_token ||
  process.env.JIRA_API_TOKEN;

const baseUrl =
  providerConfig.base_url ||
  process.env.JIRA_BASE_URL;

let connectionTest;

if (!email || !apiToken || !baseUrl) {
  const missing = [
    !email && 'JIRA_EMAIL',
    !apiToken && 'JIRA_API_TOKEN',
    !baseUrl && 'JIRA_BASE_URL',
  ].filter(Boolean);
  connectionTest = {
    skipped: true,
    reason: `missing required credential(s) or config: ${missing.join(', ')}`,
  };
} else {
  try {
    const tlsConfig = fetchUndiciMod.createTLSConfig({
      verify: connCfg.verifySsl,
      ca: connCfg.caBundlePath ? fs.readFileSync(connCfg.caBundlePath) : null,
      cert: connCfg.clientCert ? fs.readFileSync(connCfg.clientCert) : null,
      key: connCfg.clientKey ? fs.readFileSync(connCfg.clientKey) : null,
    });

    const proxy = connCfg.proxyUrl ? fetchUndiciMod.createProxy(connCfg.proxyUrl) : null;

    const client = new fetchUndiciMod.AsyncClient({
      tls: tlsConfig,
      proxy,
      trustEnv: true,
    });

    const basicCredential = Buffer.from(`${email}:${apiToken}`).toString('base64');
    const endpoint = `${baseUrl.replace(/\/$/, '')}/rest/api/3/myself`;

    const start = performance.now();
    const response = await client.get(endpoint, {
      headers: {
        'Authorization': `Basic ${basicCredential}`,
        'Accept': 'application/json',
      },
    });
    const latency = Math.round(performance.now() - start);

    const body = await response.json();
    await client.close();

    connectionTest = {
      skipped: false,
      connected: true,
      endpoint,
      status_code: response.statusCode,
      hasAccountId: typeof body === 'object' && body !== null && 'accountId' in body,
      latency_ms: latency,
      connection: {
        proxy: connCfg.proxyUrl || '(none)',
        verify_ssl: connCfg.verifySsl,
        ca_bundle: connCfg.caBundlePath || '(default)',
        has_client_cert: Boolean(connCfg.clientCert),
      },
    };
  } catch (err) {
    connectionTest = {
      skipped: false,
      connected: false,
      error: { name: err.name, message: err.message },
    };
  }
}

const summary = {
  service: 'jira-fetch-undici',
  mode: 'connection-test',
  mjs_package: '@internal/fetch-undici',
  app_yaml_initialized: true,
  connection_config: connCfg,
  connection_test: connectionTest,
  overall: connectionTest.connected ? 'PASS' : connectionTest.skipped ? 'SKIP' : 'FAIL',
};

// Redact sensitive fields before logging (CodeQL js/clear-text-logging)
const redactedSummary = { ...summary, connection_config: { ...summary.connection_config, proxyUrl: summary.connection_config.proxyUrl ? '*******' : null, clientKey: summary.connection_config.clientKey ? '*******' : null } };
console.log(JSON.stringify(redactedSummary, null, 2));
